### PR TITLE
refactor: migrate to intoto/attestations go library -> in-toto 1.0 bump 

### LINKED ISSUE #60: in-toto 1.0 support
In-toto 1.0 [has been released](https://github.com/in-toto/attestation/releases/tag/v1.0) and it includes a couple of major enhancements that seem relevant to the Chainloop Project.

### [New upstream predicates](https://github.com/in-toto/attestation/tree/main/spec/predicates#vetted-predicates) and [guidelines](https://github.com/in-toto/attestation/blob/main/docs/new_predicate_guidelines.md) available

Chainloop uses a custom predicate `chainloop.dev/attestation/v0.1`. We should re-evaluate the need of our custom predicate or use an existing one. If we still do, we should make sure it adheres to the guidelines and if it would make sense to try to upstream it.

### Attestation bundle 

The new [attestation bundle](https://github.com/in-toto/attestation/blob/main/spec/v1.0/bundle.md) can be specially handy in the context of https://github.com/chainloop-dev/chainloop/issues/53. It will allow us to bundle not only the Chainloop generated DSSE envelope but any other generated envelope. 

For extensibility in the future, it might be a good idea to default to a bundle for communication with the control-plane  

### Resource descriptor 

Currently Chainloop encodes the materials in the attestation like this. It's basically an extension of a slsa material (with digest and URI) plus some additional metadata used by the Chainloop controlplane for validation and routing (name and type)

The new [resource descriptor](https://github.com/in-toto/attestation/blob/main/spec/v1.0/resource_descriptor.md) might be a good fit to replace that custom type. The `annotations` field could contain the custom chainloop types. 

```json
 "materials": [
      {
        "material": {
          "slsa": {
            "digest": {
              "sha256": "7e5f829611edeedf36a7318f78e6914b85a8e2f90ac8628cc868b2e87204d93b"
            },
            "uri": "ghcr.io/chainloop-dev/frontend"
          }
        },
        "name": "frontend",
        "type": "CONTAINER_IMAGE"
      },
      {
        "material": {
          "slsa": {
            "digest": {
              "sha256": "092252ea069679c658a949748ca0bee9405d36630d6b460964c7225708de0bb5"
            },
            "uri": "sbom.cyclonedx.json"
          }
        },
        "name": "frontend-sbom",
        "type": "SBOM_CYCLONEDX_JSON"
      }
    ],

```

Congrats @adityasaky, @TomHennen and rest of the team for the release! Feel free to drop any thoughts or comments, always appreciated!

### PR DESCRIPTION
**Pull request by migmartri**
Attestation rendering logic refactoring to support in-toto v1.0. 

Previously, we did some work https://github.com/chainloop-dev/chainloop/pull/103 adopting `resourceDescriptors` in our custom predicate, but now, in-toto 1.0 spec is fully supported.  

In practice, this patch:

- Replaces https://github.com/in-toto/in-toto-golang in favor of https://github.com/in-toto/attestation/tree/main. We now leverage the proto-bindings implemented by @marcelamelara, @adityasaky, and the rest of the in-toto community.
- Enables annotations in the attestation subject, to enable 383

Closes #60

### CODE REVIEW

File: app/cli/internal/action/attestation_push.go [hunk 1]
```diff
@@ -118,31 +118,32 @@ func (action *AttestationPush) Run(runtimeAnnotations map[string]string) (*Attes
...
+	return attestationResult, nil
 
-	action.Logger.Debug().Msg("render completed")
-	if action.c.CraftingState.DryRun {
-		action.Logger.Info().Msg("dry-run completed, push skipped")
-		// We are done, remove the existing att state
-		if err := action.c.Reset(); err != nil {
-			return nil, err
-		}
+	// action.Logger.Debug().Msg("render completed")
```
Comment (migmartri) [inline app/cli/internal/action/attestation_push.go:123]: I'll need to fix these tests.

File: app/cli/internal/action/attestation_push.go [hunk 2]
```diff
@@ -107,6 +109,9 @@ func (action *AttestationPush) Run(runtimeAnnotations map[string]string) (*Attes
 
 	action.Logger.Debug().Msg("validation completed")
 
+	// Indicate that we are done with the attestation
+	action.c.CraftingState.Attestation.FinishedAt = timestamppb.New(time.Now())
```
Comment (migmartri) [inline app/cli/internal/action/attestation_push.go:113]: The finishedAt timestamp is now in the attestation state. This simplifies testing by making rendering idempotent.

File: app/cli/internal/action/workflow_run_describe.go [hunk 3]
```diff
@@ -26,7 +26,7 @@ import (
 	"github.com/chainloop-dev/chainloop/internal/attestation/renderer/chainloop"
 	sigs "github.com/sigstore/cosign/v2/pkg/signature"
 
-	"github.com/in-toto/in-toto-golang/in_toto"
```
Comment (migmartri) [inline app/cli/internal/action/workflow_run_describe.go:29]: moved away from `github.com/in-toto/in-toto-golang/in_toto`

File: internal/attestation/renderer/chainloop/testdata/attestation.output-2.v0.2.json [hunk 4]
```diff
@@ -0,0 +1,96 @@
+{
+  "type": "https://in-toto.io/Statement/v1",
```
Comment (migmartri) [inline internal/attestation/renderer/chainloop/testdata/attestation.output-2.v0.2.json:2]: in-toto v1! 🎉

Comment (migmartri) [inline internal/attestation/renderer/chainloop/testdata/attestation.output-2.v0.2.json:2]: in-toto v1! 🎉

File: internal/attestation/renderer/chainloop/v02.go [hunk 5]
```diff
@@ -90,21 +99,44 @@ func (r *RendererV02) Header() (*in_toto.StatementHeader, error) {
...
+
+	p := ProvenancePredicateV02{
+		ProvenancePredicateCommon: predicateCommon(r.builder, r.att),
+		Materials:                 normalizedMaterials,
+	}
+
+	// transform to structpb.Struct in a two steps process
+	// 1 - ProvenancePredicate -> json -> map[string]interface{}
+	// 2 - map[string]interface{} -> structpb.Struct
+	predicateJSON, err := json.Marshal(p)
```
Comment (migmartri) [inline internal/attestation/renderer/chainloop/v02.go:126]: a little bit of trickery to transform to structpb :(

File: internal/attestation/renderer/renderer.go [hunk 6]
```diff
@@ -43,8 +44,7 @@ type AttestationRenderer struct {
 }
 
 type r interface {
-	Header() (*in_toto.StatementHeader, error)
-	Predicate() (interface{}, error)
+	Statement() (*intoto.Statement, error)
```
Comment (migmartri) [inline internal/attestation/renderer/renderer.go:47]: the interface has been simplified now.

**Comment by migmartri** [comment]
@adityasaky, re this comment https://github.com/chainloop-dev/chainloop/issues/60#issuecomment-1540296270

I can see some deprecation warnings in the `in-toto-golang` repo but not an actual implementation using those new proto bindings. 

Am I missing anything? Is the path forward recommended by the in-toto team to adopt the bindings at `in-toto/attestation/go` still?

Thank you!

**Comment by migmartri** [comment]
> > I can see some deprecation warnings in the in-toto-golang repo but not an actual implementation using those new proto bindings.
> 
> @migmartri Adding support for the proto bindings to in-toto-golang is in-flight: [in-toto/in-toto-golang#268](https://github.com/in-toto/in-toto-golang/pull/268) We were blocked on something until last week, so I should be able to complete this PR now.

hi @marcelamelara, thanks for they information. 

So great! We are both considering using in-toto/attestation/go as the source of truth for the bindings.

It seems that in your PR you are also adding some exciting generators, I might use them in the future :)

**Review by danlishka — APPROVED** [review]
LGTM

**Pull request merged**